Add Custom Dynamic Data Entity Source sample. (#468)
Esri/arcgis-maps-sdk-flutter-samples: lib/samples/add_custom_dynamic_entity_data_source/README.md

# Add custom dynamic entity data source

Create a custom dynamic entity data source and display it using a dynamic entity layer.

![Image of add custom dynamic entity data source](add_custom_dynamic_entity_data_source.png)

## Use case

Developers can create a custom `DynamicEntityDataSource` to be able to visualize data from a variety of different feeds as dynamic entities using a `DynamicEntityLayer`. An example of this is in a mobile situational awareness app, where a custom `DynamicEntityDataSource` can be used to connect to peer-to-peer feeds in order to visualize real-time location tracks from teammates in the field.

## How to use the sample

Run the sample to view the map and the dynamic entity layer displaying the latest observation from the custom data source. Tap on a dynamic entity to view its attributes in a callout.

## How it works

Configure the custom data source:

> 1. Create a custom data source implementation of a `DynamicEntityDataSource`.

Reviewer: ```suggestion
1. Create a custom data source implementation of a `CustomDynamicEntityDataProvider `.
```

2. Override `onnLoad()` to specify the `DynamicEntityDataSourceInfo` for a given unique entity ID field and a list of `Field` objects matching the fields in the data source.
3. Override `onConnect()` to begin processing observations from the custom data source.

> 4. Loop through the observations and deserialize each observation into an `ArcGISPoint` and a `Map<String, Object>` containing the attributes.

Reviewer: ```suggestion
4. Loop through the observations and deserialize each observation into an `ArcGISPoint` and a `Map<String, dynamic>` containing the attributes.
```

5. Use `NewObservation(geometry, attributes)` to add each observation to the custom data source.

Configure the map view:

1. Create a `DynamicEntityLayer` using the custom data source implementation.

> 2. Update values in the layer's `TrackDisplayProperties` to customize the layer's appearance.

Reviewer 2: ```suggestion
2. Update values in the layer's `trackDisplayProperties` to customize the layer's appearance.
```

> 3. Set up the layer's `LabelDefinitions` to display labels for each dynamic entity.

Reviewer 2: ```suggestion
3. Set up the layer's `labelDefinitions` to display labels for each dynamic entity.
```

4. Configure the `ArcGISMapView(onTap: ...)` handler to select a dynamic entity and display the entity's attributes in a callout.

## Relevant API

* CustomDynamicEntityDataProvider
* CustomDynamicEntityDataSource
* DynamicEntity
* DynamicEntityDataSource
* DynamicEntityLayer
* LabelDefinition
* TrackDisplayProperties

## About the data

This sample uses a [JSON Lines file containing observations of marine vessels in the Pacific Northwest](https://www.arcgis.com/home/item.html?id=a8a942c228af4fac96baa78ad60f511f) hosted on ArcGIS Online .

## Additional information

> In this sample, we iterate through features in a GeoJSON file to mimic messages coming from a real-time feed. You can create a custom dyamic entity data source to process any data that contains observations which can be translated into `MapPoint` objects with associated `Dictionary<string, object>` attributes.

Reviewer: ```suggestion
In this sample, we iterate through features in a GeoJSON file to mimic messages coming from a real-time feed. You can create a custom dyamic entity data source to process any data that contains observations which can be translated into `ArcGISPoint` objects with associated `Map<String, dynamic>` attributes.
```
